src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java

```diff
@@ -206,32 +204,41 @@ public MessageImpl createMessage(boolean send, String address, MultiMap headers,
   protected <T> void sendOrPub(OutboundDeliveryContext<T> sendContext) {
     if (((ClusteredMessage) sendContext.message).getRepliedTo() != null) {
       clusteredSendReply(((ClusteredMessage) sendContext.message).getRepliedTo(), sendContext);
+    } else if (sendContext.options.isLocalOnly()) {
+      super.sendOrPub(sendContext);
     } else {
-      if (sendContext.options.isLocalOnly()) {
-        super.sendOrPub(sendContext);
-      } else if (Vertx.currentContext() != sendContext.ctx) {
-        // Current event-loop might be null when sending from non vertx thread
-        sendContext.ctx.runOnContext(v -> {
-          subs.get(sendContext.message.address(), ar -> onSubsReceived(ar, sendContext));
-        });
+      Serializer serializer = Serializer.get(sendContext.ctx, deliveryStrategy);
+      if (sendContext.message.isSend()) {
+        send(sendContext, serializer);
       } else {
-        subs.get(sendContext.message.address(), ar -> onSubsReceived(ar, sendContext));
+        publish(sendContext, serializer);
       }
     }
   }
 
-  private <T> void onSubsReceived(AsyncResult<ChoosableIterable<ClusterNodeInfo>> asyncResult, OutboundDeliveryContext<T> sendContext) {
-    if (asyncResult.succeeded()) {
-      ChoosableIterable<ClusterNodeInfo> serverIDs = asyncResult.result();
-      if (serverIDs != null && !serverIDs.isEmpty()) {
-        sendToSubs(serverIDs, sendContext);
-      } else {
-        super.sendOrPub(sendContext);
-      }
-    } else {
-      log.error("Failed to send message", asyncResult.cause());
-      sendContext.written(asyncResult.cause());
+  private <T> void send(OutboundDeliveryContext<T> sendContext, Serializer serializer) {
+    serializer.<String>queue(
+      sendContext,
+      (strategy, message, promise) -> strategy.chooseForSend(message, promise),
+      (sc, res) -> sendToNode(sc, res),
+      (sc, throwable) -> sendOrPublishFailed(sc, throwable)
+    );
+  }
+
+  private <T> void publish(OutboundDeliveryContext<T> sendContext, Serializer serializer) {
+    serializer.<Iterable<String>>queue(
+      sendContext,
+      (strategy, message, promise) -> strategy.chooseForPublish(message, promise),
+      (sc, res) -> sendToNodes(sc, res),
+      (sc, throwable) -> sendOrPublishFailed(sc, throwable)
```

Comment:
make this a field of the ClusteredEventBus so it is cached and will not instantiate a lambda on every call